src/nvim/os/env.c

```diff
@@ -346,21 +354,50 @@ void init_homedir(void)
   }
 #endif
 
-  if (var != NULL) {
 #ifdef UNIX
+  if (var == NULL) {
+    var = os_homedir();
+  }
+
+  if (var != NULL) {
     // Change to the directory and get the actual path.  This resolves
     // links.  Don't do it when we can't return.
-    if (os_dirname((char_u *)os_buf, MAXPATHL) == OK && os_chdir(os_buf) == 0) {
-      if (!os_chdir(var) && os_dirname(IObuff, IOSIZE) == OK) {
+    if (os_dirname((char_u *)os_buf, sizeof(os_buf)) == OK
+        && os_chdir(os_buf) == 0) {
+      if (!os_chdir(var) && os_dirname(IObuff, sizeof(IObuff)) == OK) {
         var = (char *)IObuff;
       }
       if (os_chdir(os_buf) != 0) {
         EMSG(_(e_prev_dir));
       }
     }
+  }
 #endif
-    homedir = xstrdup(var);
+
+  if ((var == NULL || *var == NUL)
+      && os_dirname((char_u *)os_buf, sizeof(os_buf)) == OK) {
+    var = os_buf;
+  }
+
+  homedir = xstrdup(var);
+}
+
+static char homedir_buf[MAXPATHL];
+
+char *os_homedir(void)
+{
+  uv_mutex_lock(&mutex);
+  size_t homedir_size = sizeof(homedir_buf);
+  int ret_value = uv_os_homedir(os_buf, &homedir_size);
+  if (ret_value == 0) {
+    xstrlcpy(homedir_buf, os_buf, homedir_size + 1);
+    uv_mutex_unlock(&mutex);
+    return homedir_buf;
+  } else {
+    ELOG("uv_os_homedir() failed %d: %s", ret_value, os_strerror(ret_value));
```

Comment:
Do not put `ELOG` calls in critical sections. Because it may call `os_getenv()`, `os_setenv()` internally, there is a risk of deadlock. It is better to execute `uv mutex_unlock(&​​mutex);` immediately after calling `uv_os_homedir()`.

Reply:
good catch. IOW, put the ELOG() statement after `uv_mutex_unlock`.

```
uv_mutex_unlock(&​​mutex);
...
if (r != 0) {
  ELOG(...);
}
```